collections/composers/softRemoval.js

```diff
@@ -18,48 +18,61 @@ const filterOptions = (options = {}) => {
  * @param {Mongo.Collection} collection - The collection to enhance with soft removal functionality.
  * @returns {Object} An enhanced collection with soft removal methods.
  */
-export const softRemoval = (collection) => {
-  const originalFind = collection.find.bind(collection);
-  const originalFindOneAsync = collection.findOneAsync.bind(collection);
-  const originalUpdateAsync = collection.updateAsync.bind(collection);
-  const originalRemoveAsync = collection.removeAsync.bind(collection);
-  // eslint-disable-next-line prefer-object-spread
-  return Object.assign({}, collection, {
-    find(selector, options) {
-      return originalFind(
-        {
-          ...toSelector(selector),
-          ...filterOptions(options),
-        },
-        options
-      );
-    },
-    async findOneAsync(selector, options) {
-      return originalFindOneAsync(
-        {
-          ...toSelector(selector),
-          ...filterOptions(options),
-        },
-        options
-      );
-    },
-    async removeAsync(selector, options = {}) {
-      if (options.hardRemove) {
-        return originalRemoveAsync(selector);
-      }
-      return originalUpdateAsync(
-        {
-          ...toSelector(selector),
-        },
-        {
-          $set: {
-            ...(options.$set || {}),
-            isRemoved: true,
-            removedAt: new Date(),
+export const softRemoval =
+  ({ afterRemove = ({ doc }) => doc } = {}) =>
+  (collection) => {
+    const originalFind = collection.find.bind(collection);
+    const originalFindOneAsync = collection.findOneAsync.bind(collection);
+    const originalUpdateAsync = collection.updateAsync.bind(collection);
+    const originalRemoveAsync = collection.removeAsync.bind(collection);
+    // eslint-disable-next-line prefer-object-spread
+    return Object.assign({}, collection, {
+      find(selector, options) {
+        return originalFind(
+          {
+            ...toSelector(selector),
+            ...filterOptions(options),
           },
-        },
-        { multi: true }
-      );
-    },
-  });
-};
+          options
+        );
+      },
+      async findOneAsync(selector, options) {
+        return originalFindOneAsync(
+          {
+            ...toSelector(selector),
+            ...filterOptions(options),
+          },
+          options
+        );
+      },
+      async removeAsync(
+        selector,
+        { shouldFetchFullDoc, hardRemove, ...options } = {}
+      ) {
+        const docsToRemove =
+          shouldFetchFullDoc &&
+          (await originalFind(toSelector(selector)).fetchAsync());
+
+        if (hardRemove) {
+          await originalRemoveAsync(selector);
+          return afterRemove({ doc: docsToRemove, collection, isRemove: true });
```

Comment:
Since `docsToRemove` is an array, I believe the prop in the `afterRemove` parameter should be `docs`.

Reply:
A prop to indicate to the `hardRemove` will be useful here (`isHardRemove: true`)